crates/cargo-wdk/src/actions/new/mod.rs

```diff
@@ -291,3 +291,476 @@ impl<'a> NewAction<'a> {
+        let test_new_action = TestNewAction::new(empty_path)
+            .expect_cargo_new(None, None)
+            .expect_copy_lib_rs_template(true)
+            .expect_update_cargo_toml(true, true, true);
+        let new_action = NewAction::new(
+            empty_path,
+            driver_type,
+            verbosity_level,
+            &test_new_action.mock_exec,
+            &test_new_action.mock_fs,
+        );
+        let result = new_action.run();
+        assert!(
+            matches!(result, Err(NewActionError::InvalidDriverCrateName(_))),
+            "Expected InvalidDriverCrateName error"
+        );
+    }
+
+    #[test]
+    fn when_copy_build_rs_template_fails_then_returns_filesystem_error() {
+        let path = Path::new("test_driver_fail_build_rs");
+        let driver_type = DriverType::Kmdf;
+        let verbosity_level = Verbosity::default();
+
+        // Set up mocks with failure at copy build rs template to driver project step
+        let test_new_action = TestNewAction::new(path)
+            .expect_cargo_new(None, None)
+            .expect_copy_lib_rs_template(true)
+            .expect_update_cargo_toml(true, true, true)
+            .expect_create_inx_file(true)
+            .expect_copy_build_rs_template(false); // Force failure here
+
+        let result = NewAction::new(
+            path,
+            driver_type,
+            verbosity_level,
+            &test_new_action.mock_exec,
+            &test_new_action.mock_fs,
+        )
+        .run();
+        assert!(
+            matches!(
+                result,
+                Err(NewActionError::FileSystem(FileError::WriteError(_, _)))
+            ),
+            "Expected FileSystem WriteError from copy_build_rs_template step"
+        );
+    }
+
+    #[test]
+    fn when_copy_cargo_config_fails_then_returns_filesystem_error() {
+        let path = Path::new("test_driver_fail_cargo_config");
+        let driver_type = DriverType::Kmdf;
+        let verbosity_level = Verbosity::default();
+
+        // Set up mocks with failure at copy cargo config to driver project step
+        let test_new_action = TestNewAction::new(path)
+            .expect_cargo_new(None, None)
+            .expect_copy_lib_rs_template(true)
+            .expect_update_cargo_toml(true, true, true)
+            .expect_create_inx_file(true)
+            .expect_copy_build_rs_template(true)
+            .expect_copy_cargo_config(false); // Force failure here
+
+        let result = NewAction::new(
+            path,
+            driver_type,
+            verbosity_level,
+            &test_new_action.mock_exec,
+            &test_new_action.mock_fs,
+        )
+        .run();
+        assert!(
+            matches!(
+                result,
+                Err(NewActionError::FileSystem(FileError::WriteError(_, _)))
+            ),
+            "Expected FileSystem WriteError from copy_cargo_config step"
+        );
+    }
+
+    struct TestNewAction<'a> {
```

Comment:
```suggestion
    struct TestSetup<'a> {
```

`TestSetup` is a clearer name. The problem with `TestNewAction` is that it conveys almost no meaning at all.